File: packages/js/wasm/src/InMemoryFileReader.ts
import { WRAP_MANIFEST_PATH, WRAP_MODULE_PATH } from "./constants";
import { IFileReader } from "./IFileReader";

import { Result, ResultErr, ResultOk } from "@polywrap/result";

export class InMemoryFileReader {
  static fromManifest(
    manifest: Uint8Array,
    baseFileReader: IFileReader
  ): IFileReader {
    return {
      async readFile(filePath: string): Promise<Result<Uint8Array, Error>> {
        if (filePath === WRAP_MANIFEST_PATH) {
          return ResultOk(manifest);
        } else {
          return baseFileReader.readFile(filePath);
        }
      },
    } as IFileReader;
  }

  static fromWasmModule(
    wasmModule: Uint8Array,
    baseFileReader: IFileReader
  ): IFileReader {
    return {
      async readFile(filePath: string): Promise<Result<Uint8Array, Error>> {
        if (filePath === WRAP_MODULE_PATH) {
          return ResultOk(wasmModule);
        } else {
          return baseFileReader.readFile(filePath);
        }
      },
    };
  }
  static from(
    manifest: Uint8Array,
    wasmModule: Uint8Array,
    baseFileReader?: IFileReader
  ): IFileReader {
    return {
      async readFile(filePath: string): Promise<Result<Uint8Array, Error>> {
        if (filePath === WRAP_MANIFEST_PATH) {
          return ResultOk(manifest);
        } else if (filePath === WRAP_MODULE_PATH) {
          return ResultOk(wasmModule);
        } else if (baseFileReader) {
          return baseFileReader.readFile(filePath);
        } else {
          return ResultErr(Error(`File not found at ${filePath}.`));
        }
      },
    };
  }
}

File: packages/py/polywrap-wasm/polywrap_wasm/inmemory_file_reader.py
from polywrap_core import IFileReader

from .constants import WRAP_MODULE_PATH


class InMemoryFileReader(IFileReader):
    _wasm_module: bytearray
    _base_file_reader: IFileReader

    def __init__(self, wasm_module: bytearray, base_file_reader: IFileReader):
        self._wasm_module = wasm_module
        self._base_file_reader = base_file_reader

    async def read_file(self, file_path: str) -> bytearray:
        if file_path == WRAP_MODULE_PATH:
            return self._wasm_module
        else:
            return await self._base_file_reader.read_file(file_path=file_path)
